cli/src/commands/evolog.rs

```diff
@@ -151,14 +161,35 @@ pub(crate) fn cmd_evolog(
     if !args.no_graph {
         let mut raw_output = formatter.raw()?;
         let mut graph = get_graphlog(graph_style, raw_output.as_mut());
-        for commit in commits {
-            let edges = commit
-                .predecessor_ids()
+
+        let commit_dag: Vec<Result<GraphNode<Commit>, Infallible>> = commits
+            .iter()
+            .map(|c| {
+                let edges = c
+                    .predecessors()
+                    .map(|p| GraphEdge::direct(p.unwrap().clone()))
```

Comment:
nit: this `.unwrap()` can panic. We can use `.map_ok()` and `.try_collect()` instead.
```rust
                c.predecessors()
                    .map_ok(GraphEdge::direct)
                    .try_collect()
                    .map(|edges| (c, edges)) // or .try_collect()?; Ok((c, edges)) with type annotation
```

Another idea is to construct `Vec<GraphNode<Commit>>` upfront at `topo_order_reverse_ok()`.

Reply:
To be honest this particular suggestion has me a bit lost. I tried using `map_ok` and `try_collect` but it ended up pushing the potential `BackendError` up a level, so that `commit_dag` became `Vec<Result<_, BackendError>`, and I'd need to unsafely unwrap it somewhere else (like on line 179).

What's the desired behavior if we fail to fetch predecessors? I'd assume the whole command should fail with an error message, but I'm not sure what the mechanics of making that happen are.

Reply:
Try something liek this:
```rust
        let commit_dag: Vec<GraphNode<Commit>> = commits
            .into_iter()
            .map(|c| -> Result<_, BackendError> {
                let edges = c.predecessors().map_ok(GraphEdge::direct).try_collect()?;
                Ok((c, edges))
            })
            .try_collect()?;

        // XXX this isn't actually an "iter"
        let iter_nodes = if args.reversed {
            reverse_graph(commit_dag.into_iter().map(Result::<_, Infallible>::Ok)).unwrap()
        } else {
            commit_dag
        };
```

Another option:
```rust
    let to_graph_node = |commit: Commit| -> Result<GraphNode<Commit>, BackendError> {
        let edges = commit.predecessors().map_ok(GraphEdge::direct).try_collect()?;
        Ok((commit, edges))
    };
    let commit_dag = topo_order_reverse_ok(
        [to_graph_node(start_commit)],
        |(commit, _)| commit.id().clone(),
        |(_, edges)| {
            // Predecessors don't need to follow any defined order. However in
            // practice, if there are multiple predecessors, then usually the
            // first predecessor is the previous version of the same change, and
            // the other predecessors are commits that were squashed into it. If
            // multiple commits are squashed at once, then they are usually
            // recorded in chronological order. We want to show squashed commits
            // in reverse chronological order, and we also want to show squashed
            // commits before the squash destination (since the destination's
            // subgraph may contain earlier squashed commits as well), so we
            // visit the predecessors in reverse order.
            edges.iter().rev().map(|edge| to_graph_node(edge.target.clone())).collect_vec()
        },
    )?;
```